### ISSUE TITLE
HGNResult memory management

### ISSUE THREAD

**Issue by OleguerCanal** [CampusAI/Hamiltonian-Generative-Networks#24]
Make sure we are not doing copies of the intermediate results in the HGNResult class, we want pointers.

**Comment by fedetask** [comment]
The only improvement I see in `HgnResult` is to avoid copying the reconstructed rollouts into the tensor used for computing the loss, and appending them instead to a list. Then, we would need to compute the loss as 
```python
# Let prediction be the HgnResult object containing the reconstructed rollouts
# Let self.loss be the reconstruction loss used

loss = torch.tensor(0.)
for i, reconstructed_rollout in enumerate(prediction.reconstructed_rollouts):
    loss += self.loss(reconstructed, prediction.input[i])
``` 
But I don't think this is faster.
